common/persistence/visibility_hybrid_manager.go

```diff
@@ -0,0 +1,580 @@
+			mgr, ok := v.visibilityMgrs[storeName]
+			if !ok || mgr == nil {
+				return fmt.Errorf("Visibility store manager with name %s not found", storeName)
+			}
+			return mgr.RecordWorkflowExecutionClosed(ctx, request)
+		},
+	)
+}
+
+func (v *visibilityHybridManager) RecordWorkflowExecutionUninitialized(
+	ctx context.Context,
+	request *RecordWorkflowExecutionUninitializedRequest,
+) error {
+	return v.chooseVisibilityManagerForWrite(
+		ctx,
+		func(storeName string) error {
+			mgr, ok := v.visibilityMgrs[storeName]
+			if !ok || mgr == nil {
+				return fmt.Errorf("Visibility store manager with name %s not found", storeName)
+			}
+			return mgr.RecordWorkflowExecutionUninitialized(ctx, request)
+		},
+	)
+}
+
+func (v *visibilityHybridManager) DeleteWorkflowExecution(
+	ctx context.Context,
+	request *VisibilityDeleteWorkflowExecutionRequest,
+) error {
+	return v.chooseVisibilityManagerForWrite(
+		ctx,
+		func(storeName string) error {
+			mgr, ok := v.visibilityMgrs[storeName]
+			if !ok || mgr == nil {
+				return fmt.Errorf("Visibility store manager with name %s not found", storeName)
+			}
+			return mgr.DeleteWorkflowExecution(ctx, request)
+		},
+	)
+}
+
+func (v *visibilityHybridManager) DeleteUninitializedWorkflowExecution(
+	ctx context.Context,
+	request *VisibilityDeleteWorkflowExecutionRequest,
+) error {
+	return v.chooseVisibilityManagerForWrite(
+		ctx,
+		func(storeName string) error {
+			mgr, ok := v.visibilityMgrs[storeName]
+			if !ok || mgr == nil {
+				return fmt.Errorf("Visibility store manager with name %s not found", storeName)
+			}
+			return mgr.DeleteUninitializedWorkflowExecution(ctx, request)
+		},
+	)
+}
+
+func (v *visibilityHybridManager) UpsertWorkflowExecution(
+	ctx context.Context,
+	request *UpsertWorkflowExecutionRequest,
+) error {
+	return v.chooseVisibilityManagerForWrite(
+		ctx,
+		func(storeName string) error {
+			mgr, ok := v.visibilityMgrs[storeName]
+			if !ok || mgr == nil {
+				return fmt.Errorf("Visibility store manager with name %s not found", storeName)
+			}
+			return mgr.UpsertWorkflowExecution(ctx, request)
+		},
+	)
+}
+
+func (v *visibilityHybridManager) chooseVisibilityModeForAdmin() string {
+	var modes []string
+
+	for storeName, mgr := range v.visibilityMgrs {
+		if mgr != nil {
+			modes = append(modes, storeName)
+		}
+	}
+
+	if len(modes) == 0 {
+		return "INVALID_ADMIN_MODE"
+	}
+
+	return strings.Join(modes, ",")
+}
+
+func (v *visibilityHybridManager) chooseVisibilityManagerForWrite(ctx context.Context, visFunc func(string) error) error {
+	var writeMode string
+	if v.writeVisibilityStoreName != nil {
+		writeMode = v.writeVisibilityStoreName()
+	} else {
+		key := VisibilityAdminDeletionKey("visibilityAdminDelete")
+		if value := ctx.Value(key); value != nil && value.(bool) {
+			writeMode = v.chooseVisibilityModeForAdmin()
+		}
+	}
+	modes := strings.Split(writeMode, ",")
+
+	var errors []string
+	for _, mode := range modes {
+		if mode == advancedWriteModeOff {
```

Comment:
if advancedWriteModeOff, do we still expect other modes?

Reply:
We should not, we have checks when server starts up to check if we will write to advanced visibility and determine if we need visibility_hybrid_manager.
The problem here is we get the mode from flipr, so it can be changed when we already used visibility_hybrid_manager. It can handles only write to db case which is the same as dbVisibilityManager